### PR TITLE
Option to skip CoS copying of class files when the target is already newer, such as from an external compilation

### PR DESCRIPTION
**Pull request by jglick**
Related to [#227791](https://netbeans.org/bugzilla/show_bug.cgi?id=227791) and linked issues. This is a robustness fix: in NetBeans currently, when the built-in compiler gets confused during classpath scanning and writes out broken class files due to various subtle bugs, as happens pretty frequently in my experience on big complex projects, CoS-mode execution (e.g., running of individual unit tests in quick mode) is broken because the test JVM loads bytecode full of errors (“uncompilable source code” etc.). This is irritating because the errors are almost always from classes deep in the dependency tree and unrelated to what you are actually editing and testing, yet the problem blocks you from using CoS mode and so you must fall back to much slower full builds.

With this patch, NetBeans will just trust the external build tool preferentially, based on `*.class` timestamp. Thus, if you notice some error during a CoS test case run which was caused by an IDE problem, you have a simple workaround: just do a command-line rebuild of the module containing the class mentioned in the error. You can still make edits to unrelated (source or test) files and rerun the test to pick up those changes. I have been running with this patch for over a year and it seems to work well; I no longer curse in frustration at nbjavac bugs, just note them and move on.

`java.source.base` is giving me 29 test failures after 20m (!) run time, but I picked one at random and confirmed that it was also failing in `master`, so at least that one is not a regression. When was the last time `mvn -f whatever.module test` actually worked?

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
Robustness is needed. Btw. @jlahoda provided me with a patch that works the other way around: e.g. it can adopt the existing `.class` files internally. That would be useful enhancement as well.

If this change doesn't clash with the future *copy instead of parse* enhancement, let's go for it.

**Comment by eirikbakke** [comment]
> just do a command-line rebuild of the module containing the class mentioned in the error
I assume this also works with maven builds invoked via "Clean and Build" from the project context menu, rather than from the command-line?

Thanks for working on CoS robustness! One of my big annoyances with NB9 has been when NetBeans erroneously marks source files as having errors. Sometimes a clean build has not been enough to resolve this--I've had to restart NetBeans or even delete the cache directory. If a clean build can now resolve these cases, it makes many other bugs less serious.

File: java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (emilianbold) [inline java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: You are already doing something with .sig here...

File: java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (emilianbold) [inline java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: ... what does this other thing you do with .sig here?

File: java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (jglick) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: This is the same as the original line 320. [Reduced diff](https://github.com/apache/incubator-netbeans/pull/504/files?w=1)

File: java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (jglick) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: Here we are going from, say, `/prj/src/main/java/pkg/Outer$Inner.sig` (obviously not a real path; constructed at line 847) to `/prj/src/main/java/pkg/Outer.java` which we suspect is an actual file whose timestamp we could compare to that of the corresponding `/userdir/var/cache/index/s1234/java/15/classes/pkg/Outer$Inner.sig`.

Comment (jlahoda) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: So, I first thought this patch is only copying the sig files if they are newer than the classfiles.

But if I understand it correctly, this patch goes much farther. Consider these two sources:
---A.java
public class A {

    public static void main(String[] args) {
        B.hello(1);
    }
    
}
---B.java
public class B {
    
    public static void hello(int i) {
        System.err.println("Hello!");
    }

}
---

Run A (with CoS). Then change B.java to:
---B.java
public class B {
    
    public static void hello(long i) {
        System.err.println("Hello!");
    }

}
---

Save and run (A) again:
---
run:
Exception in thread "main" java.lang.NoSuchMethodError: B.hello(I)V
	at A.main(A.java:4)
/home/lahvac/src/nb/apache-netbeans.jdk10/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:111: The following error occurred while executing this line:
/home/lahvac/src/nb/apache-netbeans.jdk10/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
---

It (seems to) work OK without this patch. Is this what we wanted? In particular, did we want to force this unconditionally to everyone (who is using CoS)? I can understand some can live with this bug, but can all our users? Aren't we exchanging a set of (undisclosed) probably fixable bugs for (one) systematic unfixable known obvious bug?

Am I missing something?

Comment (jglick) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: > I first thought this patch is only copying the sig files if they are newer than the classfiles.

No, it is skipping the `*.sig` file copy when the `*.class` is newer than the `*.java`. In other words, trust the build system over the IDE: only run IDE-generated bytecode corresponding to source files the user has actually edited since the last external build.

> Aren't we exchanging a set of (undisclosed) probably fixable bugs for (one) systematic unfixable known obvious bug?

Yes, that is one perspective. Another is that we are mitigating a set of bugs which have not in fact been fixed in the past decade or so and which had no viable workaround, for a bug with two simple workarounds: do a clean build¹; or simply make a no-op edit to `A.java` (SPACE BACKSPACE) before rerunning it.

My experience has been that the kind of change you describe (`int` → `long` without any modifications to the caller) is pretty unusual; I do not recall ever hitting such a problem in the time that I have been running with this patch, whereas the kinds of problems that were routinely breaking CoS execution and interrupting my work have disappeared. YMMV.

The question is whether this special case can also be handled without reducing the overall benefit of the patch. I am doubtful. The example you describe is that `A.class` has a newer timestamp than `A.java` but `A.sig` has a newer timestamp than either (because the IDE’s internal compiler does dependency tracking), and we would prefer to pick `A.sig` since it is calling the new method signature. However as a user you have no control over when the IDE’s internal compiler decides to update `*.sig` files, and many NetBeans users are familiar with the dreaded phenomenon of unexpected classpath scanning and bogus error badges. That is the problem: an IDE bug can spray unwanted `*.sig` file updates all over your classpath. You can hope that there have been no such updates since your last clean build, but this is pretty fragile. In fact clean builds can themselves sometimes trigger scanning, for example when you have generated source roots.

¹Or an incremental build, depending on how sophisticated your project’s build system is. For example, NB Ant-based projects work correctly in this case after you simply `compile`.

Comment (jlahoda) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: So, another example (with this patch), but this time needs a code change to make it deterministic: add:
try {Thread.sleep(3000);} catch (Throwable t) {}
as the very first line of BuildArtifactsMapperImpl's
private Boolean performUpdate(@NonNull final Context ctx) throws IOException {
(this is to ensure timestamps change)

Then create a J2SE project with two files:
```
---cos/test/CosTest.java
package cos.test;

public class CosTest {

    public static void main(String[] args) {
//        System.err.println(Dep.get());
    }
    
}
---cos/test/Dep.java
package cos.test;

public class Dep {

//    static String get() {
//        return "Hello!";
//    }
    
}
---
```

-do a clean build
-(CoS) run
-uncomment the line in CosTest, wait 5+seconds
-go to Dep, uncomment the method in Dep
-wait a few seconds, (CoS) run:
```
run:
Exception in thread "main" java.lang.RuntimeException: Uncompilable source code - Erroneous tree type: <any>
	at cos.test.CosTest.main(CosTest.java:6)
/home/lahvac/src/nb/jdk-modules/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:111: The following error occurred while executing this line:
/home/lahvac/src/nb/jdk-modules/nbbuild/testuserdir/var/cache/executor-snippets/run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
```
-clean the project (do not build)
-(CoS) run:
```
run:
Hello!
BUILD SUCCESSFUL (total time: 0 seconds)
```

I.e. does not seem to be an indexing problem, but rather a CoS copying problem.

Comment (jglick) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: > I was also digging around compile-time constants

Yes, this is another of the corner cases where the bytecode produced by a source file can change depending on changes in _other_ source files. There are some others involving inheritance, imports, etc. Again my intuition is that these cases are not important in practice compared to overall robustness, but perhaps I have just been blessed to work on projects with large Java code bases written using “interesting” design patterns and too-clever build scripts? It would be easy enough to make this patch be enabled via yet another checkbox, at the cost of UI complexity.

> nor apisupport.ant projects use it

Huh, I thought we used to use it for NB modules?

> it would be useful if bugs in the indexing would get reported

Agreed. It can get tricky to produce reports with steps to reproduce, especially in proprietary sources. `var/log/*.dump` files _could_ help; I have thousands of them, all apparently “coupling errors”, but am not sure what to _do_ with them exactly. There are also various bug reports with `AssertionError`s from (patched) javac internals, which sometimes get fixed and sometimes not.

> CoS also supports source code that does not compile

FWIW I have never used this feature myself.

> this patch complicates that significantly.

Does it? If you have uncompilable sources, clearly you introduced these changes since the last external build—or your build would have failed. So those edits should be producing new `*.java` and `*.sig` files, which CoS execution would still honor.

> CoS can be disentangled from java.source, and can be done solely in the projects, avoiding any indexing issues altogether.

Not sure what you mean by this.

Comment (jlahoda) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: >> nor apisupport.ant projects use it

> Huh, I thought we used to use it for NB modules?

So, how do I enable it?

>> CoS also supports source code that does not compile

> FWIW I have never used this feature myself.

Me neither. If we can generalize that, I think it opens paths to very significant simplifications.

>> this patch complicates that significantly.

> Does it? If you have uncompilable sources, clearly you introduced these changes since the last external build—or your build would have failed. So those edits should be producing new *.java and *.sig files, which CoS execution would still honor.

See above for a particular example.

>> CoS can be disentangled from java.source, and can be done solely in the projects, avoiding any indexing issues altogether.

> Not sure what you mean by this.

Well, if ability to run erroneous source is not needed, then there's no strong reason why CoS in projects would need to use the outcomes of indexing. So the project (Maven/Ant/...) could simply rebuild what it sees fit, by means it sees fit (e.g. using ToolsProvider, or running a simplified version of the build script, etc.). Yes, this would mean some of the sources would be compiled twice, but:
a) indexing could, I think, be significantly simplified if it didn't have to produce runnable classfiles, improving robustness and performance.
b) CoS would not depend on the un-robust (dreaded) indexing, presumably improving quality for CoS users

Comment (jglick) [reply java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java]: > -uncomment the line in CosTest, wait 5+seconds
> -go to Dep, uncomment the method in Dep

Are you saving the sources here, or letting the save be done by F6? If the latter, that does sound like a race condition here: you would want to let all newly saved sources be scanned into sig files before the run starts, which the IDE normally ensures, and perhaps that is getting broken by this change? Or if the former (manual saves in reverse order), then, well, do not do that. Either way, you can simply touch `CosTest` again to work around.

> the project (Maven/Ant/...) could simply rebuild what it sees fit, by means it sees fit (e.g. using ToolsProvider, or running a simplified version of the build script, etc.).

Ah, I get your point now. Using an external build script is not going to work well, since this adds launch overhead, and there could be many different modified sources with different build roots that need recompilation, which would in general require multiple tool executions in sequence. Also there is typically a difference between the classpath entries used by an external build (e.g., `~/.m2/repository/upstream/1.0-SNAPSHOT/upstream-1.0-SNAPSHOT.jar`) vs. the IDE’s notion (e.g., `/src/upstream/target/classes`).

So you could use `javax.tools.JavaCompiler`. Compared to using nbjavac, this is safer, since the only IDE bugs which could break CoS would involve a failure to determine the correct classpath and other compilation options—i.e., the same limitations as CoS generally. You still do need to topologically sort the compilations.

> indexing could […] be significantly simplified if it didn't have to produce runnable classfiles

Do I remember correctly that at one time `*.sig` files merely captured non-`private` member signatures, so were deliberately incomplete bytecode, and that this was changed to generating complete class files at some point? I also vaguely recall some design constraint imposed by the need to support unpacked WAR files.

**Comment by jlahoda** [comment]
On Tue, May 8, 2018 at 3:57 AM, Jesse Glick <notifications@github.com>
wrote:

> *@jglick* commented on this pull request.
> ------------------------------
>
> In java.source.base/src/org/netbeans/modules/java/source/usages/
> BuildArtifactMapperImpl.java
> <https://github.com/apache/incubator-netbeans/pull/504#discussion_r186599280>
> :
>
> > +            LOG.log(Level.FINER, "#227791: ignoring non-file-based source {0}", approximateSource);
> +            return false;
> +        }
> +        if (!target.isFile()) {
> +            LOG.log(Level.FINER, "#227791: {0} does not even exist", target);
> +            return false;
> +        }
> +        long targetLastMod = target.lastModified();
> +        File mockSrc;
> +        try {
> +            mockSrc = BaseUtilities.toFile(approximateSource.toURI());
> +        } catch (URISyntaxException x) {
> +            LOG.log(Level.FINER, "#227791: cannot convert " + approximateSource, x);
> +            return false;
> +        }
> +        File src = new File(mockSrc.getParentFile(), mockSrc.getName().replaceFirst("([$].+)*[.]sig$", ".java"));
>
> -uncomment the line in CosTest, wait 5+seconds
> -go to Dep, uncomment the method in Dep
>
> Are you saving the sources here, or letting the save be done by F6? If the
> latter, that does sound like a race condition here: you would want to let
> all newly saved sources be scanned
>

Yes, I am saving, sorry for not being explicit. The steps would be:

-do a clean build
-(CoS) run
-uncomment the line in CosTest, save CosTest, wait 5+seconds
-go to Dep, uncomment the method in Dep
-save Dep, wait a few seconds, (CoS) run


> into sig files before the run starts, which the IDE normally ensures, and
> perhaps that is getting broken by this change? Or if the former (manual
> saves in reverse order), then, well, do not do that. Either way, you can
> simply touch CosTest again to work around.
>

So the users need to do dependency tracking themselves. So far, this was
IDE's job.

**Comment by JaroslavTulach** [comment]
I am not following your discussions fully, but the picture I got goes again to the question whether NetBeans is an [IDE for DevOps](http://wiki.apidesign.org/wiki/DevOps) or normal developers that use compile on save...

**Comment by jglick** [comment]
> users need to do dependency tracking themselves

I guess? Really it just seemed reflexive that if you were bothering to explicitly save at all, you would save the upstream part which defined an API before working on, and saving, the downstream part which called it. But maybe there is some way to salvage this situation if people rely on it; I would have to study it.

At any rate, I like the idea of stripping the sig files down to, well, signatures and putting the CoS functionality into a more superficial layer. Sounds like it would be more intuitive and predictable: the IDE would just be performing the final compilation of recently modified sources not covered by the last external build. Deserves a prototype, certainly a bigger project than this PR which was intended as a stopgap measure.

**Comment by jglick** [comment]
I realized that in the case of Maven there is a way to combine CoS-like recompilation and test run (not sure about main class run) into a single tool execution: create an ephemeral POM in a temp dir listing the open projects with snapshot relationships as `<modules>`, then

    mvn resource:copy compiler:compile surefire:test -f /tmp/pom123.xml -DignoreMissingTests -Dtest=MyTest#case

or similar. Compared to using `javax.tools.JavaCompiler` this is a bit safer still, since even POM misinterpretations by the IDE would only affect editor functions, not test runs; and I think the overhead would be modest. (Maven CoS already runs `mvn`.)

TBD what is best for (managed or freeform) Ant- or Gradle-based builds, but this reinforces the notion that the project type can keep CoS functionality closer to the build system.

**Comment by geertjanw** [comment]
Any reason not to merge this? Or what is the current status of this and the way forward?

**Comment by jglick** [comment]
While I continue to find the behavior in this patch preferable on balance to the stock behavior, it is sufficiently controversial that it should just be considered on hold until I can find time to spend on a more ambitious alternate approach. I did not see an “experimental” label or similar, so just using the closest applicable label.

**Comment by eirikbakke** [comment]
I have been running with this patch for a couple of weeks now. I *think* it has removed a lot of previously very annoying errors, or at least saved me from doing a complete clean build every time they occur ("Annotation: An error occurred during parsing", "Absent Code attribute in method", "The bytes do not represent a valid class" etc.)

Does this patch impact "Apply Code Changes" in any way? (That feature has always been very flaky... yet it is so incredibly useful that I use it anyway.)

**Comment by jglick** [comment]
> Does this patch impact "Apply Code Changes" in any way?

Not sure. I do use that feature (and also find it flaky).

**Comment by eirikbakke** [comment]
It would be really nice to get this patch integrated. It makes a huge difference. Here's my experience:

After upgrading from NetBeans 8 to NetBeans 9, my multi-module NetBeans platform maven project would constantly need to be recompiled after errors like these:
* java.lang.ClassFormatError: Duplicate method name&signature in class file
* ClassFormatError: Absent Code attribute in method that is not native or abstract
* NoClassDefFoundError: Could not initialize class

I then applied this patch in September, on top of NetBeans 9, and had no more problems with these errors.

When NetBeans 10 was released, I started using it _without_ the patch for two weeks. The errors above started appearing daily again. It was a huge annoyance. I'm have now applied the patch again, and the errors are gone.

(More usage background: I always use Compile-on-Save, and often also Apply Code Changes in the debugger. In my own project, the "start state" is always a clean maven build--but with the patch applied, I can go for several days without needing to do a clean build again. And when I do run a clean build, it's usually because I need to for an unrelated reason.)

> While I continue to find the behavior in this patch preferable on balance to the stock behavior, it is sufficiently controversial that it should just be considered on hold until I can find time to spend on a more ambitious alternate approach.

What is the main "controversial" aspect of this patch? Are there downsides to it?

**Comment by geertjanw** [comment]
Well, there are conflicting files now. Once they're fixed, since you're a committer, you should be able to commit this.

**Comment by eirikbakke** [comment]
The conflicts are just due to the java.source.base and maven modules being moved into the top-level "java" folder.

I've tested the patch, but not reviewed the actual code (it would require a cup of coffee and a spare hour sometime--I'm not familiar with CoS inner workings). If someone else already did review this, they'd be in a better position than me in determining whether it's safe to apply the patch or not.

**Comment by eirikbakke** [comment]
I have now started using NetBeans 11 as my daily development IDE. As before, without this patch applied, I get many of the previously mentioned errors (ClassFormatError etc.). When I apply this patch, those errors go away.

Unfortunately, even with this patch applied, I now very frequently have other problems with Compile-on-Save, e.g. symbols that are not indexed (for "Find Usages" or "Go to Symbol"), red error badges on files that won't go away unless I turn CoS off and do a clean build etc. I am imagining that these latest errors were not as frequent in NetBeans 10 (with the patch in this PR applied).

I see that Jaroslav's patch https://github.com/apache/netbeans/pull/1123 was merged into NetBeans 11. How does that patch relate to the one in this PR? Is it meaningful to apply them both in the same build?

**Comment by jglick** [comment]
> even with this patch applied, I now very frequently have other problems with Compile-on-Save, e.g. symbols that are not indexed (for "Find Usages" or "Go to Symbol"), red error badges on files that won't go away

Well there have been various bugs since the inception of the Java parser. This PR merely attempts to make them not be fatal to CoS execution.

I had not seen #1123 until now, though I suppose it grew out of https://github.com/apache/netbeans/pull/504#issuecomment-381851991. Clearly it is related, but I think independent. This (#504) patch is applicable to any Java project, not a specific build technology, and should only affect the behavior of CoS execution. #1123 actually switches what the editor considers to be the authoritative source of dependencies for various purposes. If I understand it correctly, it would be similar to what for example the Maven project type already did when it detects that the Shade plugin is in use: bypass the normal binary → source linkage and just trust the build product instead.

Much as I loathe to do so, I wonder if this is important enough to integrate as a user option (off by default). I recall that there used to be some GUI in the global **Options** dialog to select different modes of CoS, but I cannot find it now. There is just the per-project **Compile on Save** checkbox.

**Comment by eirikbakke** [comment]
I am tempted to try to dive into the CoS logic and see if I can help fix, or at least make reproducible, some of the bugs that occur intermittently in my own workflow.

The problem is, if I'm running NetBeans with non-standard CoS logic enabled, it will be hard for me to understand whether certain behaviors apply to the standard NetBeans distro, or whether they are caused by the patch. Thus debugging becomes harder--doubling the testing surface, so to speak. It would be a lot easier if CoS worked in one "official" way.

At the moment, NetBeans is hard to work with without this patch enabled, and so any debugging efforts I do now would apply to "NetBeans-with-Jesse's-patch-applied" rather than the vanilla NetBeans 11.0. And I will need to use NetBeans as a working IDE over several days in order to start narrowing down these intermittent bugs.

**Comment by jglick** [comment]
There are certain limitations that are just going to be inherent in the mode proposed in this patch. For example, if you refer to a nonexistent method in a caller class, save it, then define the method in the API class, then run a test, you will get an error. (Worked around by just going back to the caller and resaving it.) But these should be pretty deterministic. I suspect that the intermittent bugs you are talking about are not bugs in CoS per se, but general bugs in the Java parser, which this patch would not affect except insofar as it may prevent them from breaking CoS execution.

Certainly if you can track down any of these with reproducible test cases, the NetBeans community would be graceful. I do see obvious bugs in the form of uncaught exceptions from javac from time to time, but historically just reporting the stack trace has not always led to a resolution: the response was often that nothing could be done without a test case. In principle the reaction should be to figure out what critical diagnostic information was missing from the stack trace and improve the code somewhere in that stack to do better next time, whether with a richer exception message, earlier assertions, etc.

The same may apply to situations which clearly represent a bug in the product, such as the editor tab displaying an error badge but the editor pane not showing any specific errors. One thing that I have long wished for is a simple UI gesture to tell the IDE to throw out all its Java parser caches for a given source root and start over—sometimes deleting the `var/cache/` directory and restarting NetBeans fixes a problem, but this is rather drastic as it forces everything to be rebuilt.

**Comment by JaroslavTulach** [comment]
> I had not seen #1123 until now, though I suppose it grew out of [#504 (comment)](https://github.com/apache/netbeans/pull/504#issuecomment-381851991).

True.

>  bypass the normal binary → source linkage and just trust the build product instead.

Right.

**Comment by eirikbakke** [comment]
> One thing that I have long wished for is a simple UI gesture to tell the IDE to throw out all its Java parser caches for a given source root and start over

I'd like to start working on this feature. Is resetting the parser cache a trivial method call somewhere, or is there more to it that I need to investigate?

(Things like "Find Usages" not working properly are driving me mad... I rely on it heavily when reviewing my own code. Currently, "Find Usages" (and "Go To Symbol", and presumably "Refactor Rename" etc.) will silently omit results at random, only to display them again if I go to the Java file in question, insert a space, and invoke Save to force CoS. Again, this is with the patch in this PR applied, on top of NetBeans 11.0.)

**Comment by lkishalmi** [comment]
@eirikbakke How this patch serves? (I guess you run your own custom built NetBeans). Also to @jglick @JaroslavTulach  and @jlahoda shall this one be closed or integrated after all? Honestly, I have not read all the comments you've made on this.

**Comment by eirikbakke** [comment]
@lkishalmi I have used this patch for two years. Every time a new NetBeans release comes around, I first try developing for some hours without the patch. Compile-on-Save is completely unusable without it on my particular project (which is a multi-module NetBeans Platform maven project).

I'm not sure what the best way forward is, since according to @jglick's comment above, this patch cannot just be applied in every case.

Even with this patch, the NetBeans index still frequently gets corrupted, with Find Usages, Go to Symbol, Refactor etc. not finding every case, and error badges showing for files without errors. Existing NetBeans, deleting the "index" directory in the NetBeans cache directory, starting NetBeans again, and waiting for a complete reindex is now unfortunately part of my daily routine.

**Comment by jglick** [comment]
I modified this patch to turn off the new behavior by default, enabled only via system property. That should make it safer to merge. Whether this should be promoted to a GUI option is another question.

**Comment by eirikbakke** [comment]
Good call with the system property. That prevents the patch from being lost, and shows more clearly what the difference in behavior is.

**Comment by jglick** [comment]
@eirikbakke if you can spare a bit of time, please try running with this patch with and without `-J-Dorg.netbeans.modules.java.source.usages.BuildArtifactMapperImpl.COMPARE_TIMESTAMPS=true` to verify that it continues to solve your CoS issues. (Yes the index can still get corrupted, leading to breakage of editor features.)

**Comment by eirikbakke** [comment]
Will do.

**Comment by eirikbakke** [comment]
Sorry for the delay... I'll be testing this patch in my working IDE for a few days now, with COMPARE_TIMESTAMPS first disabled and then enabled, to verify the latest behavior.

File: java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (eirikbakke) [inline java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java:874]: What's the rule for which files need to be added to updatedFiles? Who is listening for changes here? If the reason that the file was not copied is that it changed externally, is that something a listener somewhere should be notified of?

**Comment by eirikbakke** [comment]
@jglick could you remind me how to configure NetBeans to show the various FINE log messages in BuildArtifactMapperImpl? I want to convince myself that I am successfully managing to turn this feature on and off...

File: java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (jglick) [reply java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java:874]: I do not think so. The idea here is that a build created/updated, say, `target/classes/pkg/Clazz.class` so it now has a newer timestamp than `Clazz.sig` and we should not clobber it with an older and possibly not-quite-correct version. There is nothing to listen to. If the next thing that happens is another external build, the timestamp will be even newer, and so the next CoS run will again trust the `Clazz.class`. If on the other hand the user edits & saves `Clazz.java`, causing `Clazz.sig` to be updated, and next does a CoS run, then we will overwrite `Clazz.class` with the contents of `Clazz.sig`.

**Comment by jglick** [comment]
> how to configure NetBeans to show the various FINE log messages in BuildArtifactMapperImpl

I used the `org.netbeans.modules.logmanagement` contrib module to display logs in the **Services** tab. There are probably other ways.

File: java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (eirikbakke) [reply java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java:874]: Makes sense.

**Comment by eirikbakke** [comment]
Can confirm that the patch still works, and that it can be turned on or off. I now also finally understand what it does...

With the patch included in the IDE but without the COMPARE_TIMESTAMPS setting, I pretty quickly get the old "Absent Code attribute in method that is not native or abstract" exceptions that I've become used to in the past three years of NetBeans releases (reported on https://issues.apache.org/jira/browse/NETBEANS-430 ). When enabled (with "-J-Dorg.netbeans.modules.java.source.usages.BuildArtifactMapperImpl.COMPARE_TIMESTAMPS=true"), these exceptions go away except when the error is in the same class as is being edited. I've also examined the log messages to confirm that the functionality can be turned on and off.

Also looked at a diff between the old and the new version of the patch, and confirmed that it's just cleanup and the system property test.

(Figured the log part out... added "-J-Dorg.netbeans.modules.java.source.usages.BuildArtifactMapperImpl.level=100" to netbeans_default_options in netbeans.conf to get the messages showing.)

I think this patch should be safe to merge.

Thanks for this!

**Comment by lkishalmi** [comment]
It still can go into 12.2 if you able to rebase it on delivery branch by tomorrow EOB (PDT)

**Comment by lkishalmi** [comment]
oh and document the system property in the arch.xml

**Comment by eirikbakke** [comment]
No need to rush this patch into 12.2 as far as I'm concerned; it's a feature that's off by default and which has to be enabled by a command-line flag. It's just good to get it merged eventually so that the code isn't lost.

File: java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java
Comment (lkishalmi) [inline java/java.source.base/src/org/netbeans/modules/java/source/usages/BuildArtifactMapperImpl.java:102]: Let's document tis one in arch.xml

**Review by lkishalmi — DISMISSED** [review]

File: java/java.source.base/arch.xml
Comment (eirikbakke) [inline java/java.source.base/arch.xml:813]: Some funny quotes characters here (“ and ” instead of ", and ’ instead of '). Might upset some tools...

Comment (jglick) [reply java/java.source.base/arch.xml:813]: https://github.com/apache/netbeans/blob/51697f5bd782e14c563e35eebf9bbe5dd202d037/java/java.source.base/arch.xml#L1

**Review by lkishalmi — APPROVED** [review]

File: java/java.source.base/arch.xml
Comment (eirikbakke) [reply java/java.source.base/arch.xml:813]: I was doing a build on Windows today, and got the following amusing error :-)

```
BUILD FAILED
C:\Users\ebakke\ZRoot\nbsrc\incubator-netbeans\nbbuild\build.xml:395: The following error occurred while executing this line:
C:\Users\ebakke\ZRoot\nbsrc\incubator-netbeans\nbbuild\javadoctools\build.xml:48: The following error occurred while executing this line:
C:\Users\ebakke\ZRoot\nbsrc\incubator-netbeans\nbbuild\javadoctools\build.xml:149: javax.xml.transform.TransformerException: javax.xml.transform.TransformerException: com.sun.org.apache.xml.internal.utils.WrappedRuntimeException: Invalid byte 3 of 3-byte UTF-8 sequence.
```

Examining build.xml:149, it points to the file nbbuild/build/javadoc/alldatas.xml, which indeed appears to have a malformed UTF-8 character somewhere.

Doing some further investigation...

```
ebakke@EBLEN:/z/nbsrc/incubator-netbeans/nbbuild/build/javadoc$ diff alldatas1.xml alldatas2.xml
5139c5139
<       when running in “compile-on-save? mode
---
>       when running in “compile-on-save�? mode
```

So these odd quotes did indeed cause some trouble down the line...

Should I open a PR to fix it?

Comment (eirikbakke) [reply java/java.source.base/arch.xml:813]: Added a PR for this at https://github.com/apache/netbeans/pull/2934

**Pull request merged**
